### ISSUE TITLE
HttpWebRequest.GetRequestStream not allowing payload on GET requests

### ISSUE THREAD

**Issue by thiezn** [dotnet/corefx#11211]
Issue initially reported on PowerShell/PowerShell#2054

Making a HTTP GET requests from PowerShell using either Invoke-WebRequest or Invoke-RestMethod with a content-body throws an exception. [HTTP RFC 7231](https://tools.ietf.org/html/rfc7231#section-4.3.1) describes that a content-body is allowed but some webservers might reject the request.

>    A payload within a GET request message has no defined semantics;
   sending a payload body on a GET request might cause some existing
   implementations to reject the request.

This is making it difficult to interface with various RESTful API's like InfoBlox and ElasticSearch that rely on JSON payloads for retrieving data.

The Invoke-WebMethod uses the .NET method [HttpWebRequest.GetRequestStream](https://msdn.microsoft.com/en-us/library/d4cek6cc(v=vs.110).aspx) and raises a ProtocolViolationException.

Steps to reproduce
------------------
$uri = 'http://www.microsoft.com'
$payload = @{Microsoft='loves linux'}
$result = Invoke-WebRequest -Uri $uri -Body ($payload | ConvertTo-Json) -ContentType 'application/json' -Method Get

Expected behavior
-----------------
HTTP OK

Actual behavior
---------------
Invoke-WebRequest : Cannot send a content-body with this verb-type.

**Comment by benaadams** [comment]
Do InfoBlox and ElasticSearch not work if you change the -Method to Post?

**Comment by jasonwilliams200OK** [comment]
I think this is by design: HttpWebRequest, like HttpClient, ignores the Body with GET request (and actually throws when you set the body in C#). See https://github.com/dotnet/corefx/blob/a67304dd046b14b81e4b15520c800aaa6055472a/src/System.Net.Requests/src/System/Net/HttpWebRequest.cs#L298-L306. Note the comment about RFC compliance; one of the alias of Invoke-WebRequest cmdlet is `curl` (to wit: `help iwr`), which is utility on Unix platforms that allows us to set body (`-d` switch) with GET.

So by @benaadams's advice, `-Method Post` is the way to go. :)

**Comment by davidsh** [comment]
We even have a test to verify behavior:

```c#
        [Theory, MemberData(nameof(EchoServers))]
        public void BeginGetRequestStream_UseGETVerb_ThrowsProtocolViolationException(Uri remoteServer)
        {
            HttpWebRequest request = WebRequest.CreateHttp(remoteServer);
            Assert.Throws<ProtocolViolationException>(() =>
            {
                request.BeginGetRequestStream(null, null);
            });
        }
```

**Comment by thiezn** [comment]
Launching the request with -Method Post  won't work as that will do a HTTP POST request instead of a GET request. The API's in question require a GET request. 

Indeed very interesting comment on the .NET code @jasonwilliams200OK jasonwilliams200OK points out.
 
> Match Desktop behavior: prevent someone from getting a request stream 
    if the protocol verb/method doesn't support it. Note that this is not 
    entirely compliant RFC2616 for the aforementioned compatibility reasons. 

So this indeed confirms the code is not compliant with the HTTP RFC, just not sure what compatibility reasons they are pointing to here? To me it seems it would be beneficial for the HttpWebRequest API to allow this, especially with the current emphasis in the development world on interacting through RESTful APIs.

**Comment by benaadams** [comment]
InfoBlox's [api examples](https://community.infoblox.com/t5/API-Integration/The-definitive-list-of-REST-examples/td-p/1214) don't seem to suggest there are any GET requests with a body

ElasticSearch [examples suggest](https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-multi-get.html) using `curl -d` for "Get"'s that have a body and that changes the method to [POST rather than GET](https://github.com/curl/curl/blob/master/docs/MANUAL#L289-L299)

**Comment by jasonwilliams200OK** [comment]
Another way to send body parameters with GET:

http://stackoverflow.com/a/21326644

```sh
curl -i -H "Accept: application/json" "server:5050/a/c/getName{"param0":"pradeep"}"
```

**Comment by benaadams** [comment]
@jasonwilliams200OK I believe that's just changing the header `Accept` to `application/json` so the response returns json, dropping the "GET" method override from the question and outputting the response headers `-i`

**Comment by jasonwilliams200OK** [comment]
@benaadams, yup it turned out BODY doesn't make sense in GET request after all: http://stackoverflow.com/a/983458

>
Yes, you can send a request body with GET but it should not have any meaning. If you give it meaning by parsing it on the server and changing your response based on its contents, then you are ignoring this recommendation in the [HTTP/1.1 spec, section 4.3](https://tools.ietf.org/html/rfc2616#section-4.3):
>>
[...] if the request method does not include defined semantics for an entity-body, then the message-body [SHOULD](https://www.ietf.org/rfc/rfc2119.txt) be ignored when handling the request.


So the only difference is System.Net's impl throws an exception instead of mildly warn the request originator (or not warning at all). Either way, the BODY in GET request should be ignored by the std conforming server.

**Comment by davidsh** [comment]
While `HttpWebRequest` API has this limitation of not allowing a request body for a GET request, `System.Net.Http.HttpClient` does not. So, you might want to use that API instead.

`HttpWebRequest` API is considered a legacy API.  We preserve current Desktop behavior whenever possible.

**Comment by thiezn** [comment]
@jasonwilliams200OK showing a warning would indeed be better in my view. Having the flexibility in System.Net  would allow you to use .NET/powershell to still communicate with these kinds of implementations. Best thing would be for HTTP servers/api's should adhere fully to the RFC specifications but unfortunately this isn't the case. 

@benaadams Looking at your example on InfoBlox it seems you are right it's possible without using a payload. Their API reference guide mainly focusses on using payloads in GET requests instead of handling arguments in the URL. 

Using curl (or for instance python) it's indeed possible but would like to do this from powershell since thats the language my company uses.

**Comment by benaadams** [comment]
@thiezn no I mean when there is a body they explicitly change to a POST e.g.
```
Search for a subnet

 GET /wapi/v1.2/network?network=1.1.1.0/24

Get Next Available address from a subnet
( you need the _ref to the subnet first )

 POST /wapi/v1.2/network/ZG5zLm5ldHdvcmsuMC8yNC83?_function=next_available_ip
 Content-Type: application/json

  {"num": 1}

Get the next 5 IP addresses
( you need the _ref to the subnet first )

 POST /wapi/v1.2/network/ZG5zLm5ldHdvcmsuMC8yNC83?_function=next_available_ip
 Content-Type: application/json

  {"num": 5}
```

**Comment by jasonwilliams200OK** [comment]
@davidsh, I thought HttpClient also throws in this case (at least on net451?). Therefore I have this kind of code in my unified calls dispatcher:

````c#
            var req = new HttpRequestMessage
            {
                RequestUri = new Uri(string.Format(_url, uri)),
                Method = method
            };

            req.Headers.Authorization = GetJwtAuthorizationHeader(claims);

            if (method != HttpMethod.GET && content != null)
            {
                req.Content = content;
            }

            await _httpClient.SendAsync(req) // probably this throws exception if GET has body?
                    .ContinueWith(async taskWithMessage =>...
````

**Comment by thiezn** [comment]
Thanks @davidsh, I wasn't aware of system.Net.Http.HttpClient so will see if I can use that.

@benaadams The post examples you refer to are a special case as that launches a function on the InfoBlox appliance to find a free IP address. We already use the GET request with a payload to query InfoBlox from several python scripts successfully but yes it seems you can use url encoding to query infoblox with GET requests.

**Comment by weitzhandler** [comment]
So is there a way to invoke a GET request providing it with a complex object using `HttpClient` and `HttpRequestMessage`?

I tried the following:

```c#
public void DoThing<TContent>(TContent content, Uri uri)
{
      var request = new HttpRequestMessage(HttpMethod.Get, uri) 
      {
        Content = new ObjectContent<TContent>(content, new JsonMediaTypeFormatter());
      };
      using (var response = await SendAsync(request)) /**/;
}
```

I'm getting the following `ProtocolViolationException` exception: _Cannot send a content-body with this verb-type_.

I guess it's by design, is there a way around it, I have a server that has a GET action accepting a complex graph and computes a result value out of it, so it's really a GET, not a POST. Besides I can't really change the server code.

And BTW @ghost,
How do you format GitHub code like you did [above](https://github.com/dotnet/corefx/issues/11211#issuecomment-243002982)?

**Comment by ramon-garcia** [comment]
Hello,

This makes it imposible to use Elasticsearch search api.

The API requires GET requests with body. 

curl -XGET http://server:9200/index/_search -d 'json body req' works.
with POST request (either -XPOST, or .NET Webrequest with POST method) it replies
"405 method not allowed".
